pulp_container/app/registry_api.py

```diff
@@ -475,6 +480,138 @@ def get_queryset(self, *args, **kwargs):
         return (public_repositories | accessible_repositories).distinct()
 
 
+class FlatpakIndexDynamicView(APIView):
+    """
+    Handles requests to the /index/dynamic endpoint
+    """
+
+    authentication_classes = []
+    permission_classes = []
+
+    def get(self, request):
+        req_repositories = None
+        req_tags = None
+        req_oss = None
+        req_architectures = None
+        req_annotations = False
+        req_label_exists = set()
+        req_label_values = {}
+        for key, values in request.query_params.lists():
+            if key == "repository":
+                req_repositories = values
+            elif key == "tag":
+                req_tags = values
+            elif key == "os":
+                req_oss = values
+            elif key == "architecture":
+                req_architectures = values
+            elif key.startswith("annotation:"):
+                if key.endswith(":exists") and any(v != "1" for v in values):
+                    raise ParseError()
+                req_annotations = True
+            elif key.startswith("label:"):
+                if key.endswith(":exists"):
+                    if any(v != "1" for v in values):
+                        raise ParseError()
+                    label = key[len("label:") : len(key) - len(":exists")]
```

Comment:
you can do this:
```
In [1]: x='label:<label>:exists'

In [2]: x.split(":")
Out[2]: ['label', '<label>', 'exists']

In [4]: x.split(":")[1]
Out[4]: '<label>'
```

Reply:
But that does not work when `<label>` contains colons.  (<https://github.com/owtaylor/flagstate/blob/master/docs/protocol.md> is vague on what those labels can be; and there would of course be an ambiguity that `label:foo:exists=1` is interpreted as a check whether label `foo` exists, rather than that label `foo:exists` has the value `1`, but I think that is more a shortcoming of the protocol as defined there, rather than an indication that labels may not contain colons.  The closest we get to an actual definition of what those labels can be is probably <https://github.com/opencontainers/image-spec/blob/main/config.md> "Labels object, OPTIONAL: The field contains arbitrary metadata for the container. This property MUST use the [annotation rules](https://github.com/opencontainers/image-spec/blob/main/annotations.md#rules).", where the referenced rules merely state that "Keys SHOULD be named using a reverse domain notation - e.g. com.example.myKey", but does not rule out colons.)

Reply:
hmmm to be honest based on generic guidelines, I  never encountered labels/annotation with colons https://docs.docker.com/config/labels-custom-metadata/#key-format-recommendations

However , you are right, these are unwritten conventions and if someone really wants to put a colon into a label or annotation nothing will prevent him from doing so:
```
 podman build -t scriptcamp/nginx . --label "dynamic:label2=value1"  --annotation "dynamic:label1=value1"
STEP 1/3: FROM nginx:alpine
STEP 2/3: COPY index.html /usr/share/nginx/html/index.html
--> Using cache 696dde108fecd53a8aa0145f184dcad3f175a1f993c597b5f4d28909ceadd761
--> 696dde108fe
STEP 3/3: LABEL "dynamic:label2"="value1"
COMMIT scriptcamp/nginx
--> 1c3a9e454b7
Successfully tagged localhost/scriptcamp/nginx:latest
1c3a9e454b7e0aa1a0a6e8061af166bb813a46fd2f9d33c6fda857f3cae7c273

$ podman inspect scriptcamp/nginx:latest | jq -r .[].Annotations
{
  "dynamic:label1": "value1",
  "org.opencontainers.image.base.digest": "sha256:3c98d55fa19047f2f69159a0543661848738f64f637bb6e60c95c0ce0a2c081a",
  "org.opencontainers.image.base.name": "docker.io/library/nginx:alpine"
}
(pulp) [vagrant@pulp3-source-fedora36 nginx-image]$ podman inspect scriptcamp/nginx:latest | jq -r .[].Config.Labels
{
  "dynamic:label2": "value1",
  "io.buildah.version": "1.28.0",
  "maintainer": "NGINX Docker Maintainers <docker-maint@nginx.com>"
}

```